# ML-02 — Research Question and Provisional Lane

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/SriKeerthi16-b/Machine-Learning/blob/main/work/notebooks/w01_research_question.ipynb?flush_cache=true)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

In [17]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


## 1. My lane and why

### Provisional Lane: Refresh / Content Opportunity Scoring

I am choosing the **Refresh / Content Opportunity Scoring** lane.

The goal is to identify which pages should be reviewed first for a possible content refresh, expansion, protection, pruning, or monitoring action.

I chose this lane because the starter dataset contains page-level information about search visibility, content age, freshness, position, clicks, impressions, CTR, and engagement. These signals provide information that can potentially help prioritize pages for human review.

My goal is not simply to train a model and maximize a metric. The goal is to determine whether a data-driven ranking can help a team decide **which pages deserve attention first when review resources are limited**.

This is a provisional choice. I can change the lane later if the experiments during the next few weeks show that another question is better supported by the available data.


## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

In [18]:
# This cell is for CODE (numbers, a query, a check).
# Write your text answer in the cell ABOVE this one — typing sentences here breaks Run All.


### Search question

Which pages should be reviewed first for a potential content refresh or another appropriate action, based on observable search and engagement signals?

### Unit of analysis

The unit of analysis is a **page**. Each row represents a page and contains observable signals about its search performance and content characteristics.

### Decision

The decision is **which pages should receive human review first**.

### Action

A reviewer could use the ranking to decide whether a page should be refreshed, expanded, investigated for a possible CTR issue, protected, monitored, or considered for pruning or consolidation when supported by evidence.

The model would support the decision rather than automatically making the content change.

### Cost of a wrong recommendation

A false positive could cause a team to spend limited time reviewing a page that does not need attention.

A false negative could cause a page that needs attention to be missed or reviewed later than it should be.

Therefore, the system should be treated as decision support rather than an automatic content-changing system.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [19]:
import os
import subprocess
import pandas as pd
import numpy as np

# Go to /content
os.chdir("/content")

# Clone the starter repository if needed
REPO_URL = "https://github.com/flyrank-bih/flyrank-ml-internship-starter"
REPO_DIR = "flyrank-ml-internship-starter"

if not os.path.isdir(REPO_DIR):
    subprocess.run(
        ["git", "clone", "--depth", "1", REPO_URL, REPO_DIR],
        check=True
    )

# Go into the repository
os.chdir(REPO_DIR)

# Load the dataset
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")

# Create the declining label
df["is_declining_label"] = (
    df["trend_direction"]
    .str.lower()
    .eq("down")
    .astype(int)
)

# Create the hand-rule score
stale = (df["days_since_last_update"] >= 180).astype(int)
visible = (df["impressions_90d"] >= 500).astype(int)

df["hand_rule_score"] = (
    stale * visible * df["impressions_90d"]
)

# Precision@K function
def precision_at_k(scores, labels, k):
    order = np.argsort(-np.asarray(scores))
    topk = np.asarray(labels)[order[:k]]
    return topk.mean()

# Calculate Precision@50
precision_50 = precision_at_k(
    df["hand_rule_score"],
    df["is_declining_label"],
    50
)

# Display results
print("Number of pages:", len(df))
print("Number of columns:", df.shape[1])
print("Declining rate:", round(df["is_declining_label"].mean(), 3))
print("Hand rule Precision@50:", round(precision_50, 3))

Number of pages: 30000
Number of columns: 46
Declining rate: 0.542
Hand rule Precision@50: 0.68


### Quick look at the starter data

I loaded the starter dataset to check whether the Refresh / Content Opportunity Scoring question is worth investigating.

The dataset contains **30,000 pages**.

Using the starter definition of a declining page, approximately **54.2% of pages are classified as declining**.

As a simple baseline, the starter hand rule achieved **Precision@50 = 0.680** in the initial experiment.

These numbers make the problem worth investigating because there are many pages to prioritize and the data already provides observable signals that can be used to rank pages for review. The next question is whether additional signals and ML can provide useful improvement over a simple rule while remaining interpretable and generalizable.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

### What I can claim

- The starter dataset contains **30,000 page-level observations**.
- Approximately **54.2% of pages are classified as declining** using the starter label.
- The dataset contains observable search, content, freshness, and engagement signals.
- The simple starter hand rule achieved **Precision@50 = 0.680** in the initial experiment.
- ML can potentially help rank pages for human review.

### What I cannot claim

I cannot claim that a page will definitely improve after being refreshed.

I cannot claim that a feature causes a change in search performance simply because it is associated with the outcome.

I cannot claim that the starter Precision@50 result will generalize to new clients.

I cannot claim that ML is automatically better than a simple rule.

I also need to avoid leakage by ensuring that features do not contain information about the outcome itself.

The model should therefore be treated as **decision-support**, rather than as a system that predicts or controls Google's ranking decisions.


## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.

### My self-check

- Lane chosen: Refresh / Content Opportunity Scoring
- Unit of analysis: Page
- Decision: Which pages should be reviewed first?
- Action: Review, refresh, expand, protect, monitor, or investigate
- Real numbers shown: 30,000 pages, 54.2% declining, Precision@50 = 0.680
- The goal is decision support, not simply model training.
- Claims are limited to what the data can support.
- No client names, URLs, or private queries are included.